src/lib.rs

```diff
@@ -972,15 +774,54 @@ fn debug_executor(executor: &Executor<'_>, name: &str, f: &mut fmt::Formatter<'_
 /// The local queue *must* be flushed, and all pending runnables
 /// rescheduled onto the global queue when a runner is dropped.
 struct LocalQueue {
+    /// Queue of concurrent tasks.
     queue: ConcurrentQueue<Runnable>,
-    waker: AtomicWaker,
+
+    /// Tickers waiting on an event from this queue.
+    waiters: Event,
+
+    /// Number of tickers waiting on this queue.
+    tickers: AtomicUsize,
 }
 
 impl Default for LocalQueue {
     fn default() -> Self {
         Self {
             queue: ConcurrentQueue::bounded(512),
-            waker: AtomicWaker::new(),
+            waiters: Event::new(),
+            tickers: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl LocalQueue {
+    /// Indicate that we are now waiting on this queue.
+    fn start_ticking(&self) {
+        // Relaxed ordering is fine here.
+        let old_tickers = self.tickers.fetch_add(1, Ordering::Relaxed);
+        if old_tickers > isize::MAX as usize {
+            panic!("too many tickers waiting on one thread");
+        }
+    }
+
+    /// Indicate that we are no longer waiting on this queue.
+    #[inline]
+    fn stop_ticking(&self, state: &State) {
+        if self.tickers.fetch_sub(1, Ordering::Release) == 1 {
+            // Make sure everyone knows we're about to release tasks.
+            std::sync::atomic::fence(Ordering::Acquire);
+
+            // Drain any tasks.
+            self.drain_tasks(state);
+        }
+    }
+
+    /// Drain all tasks from this queue.
+    #[cold]
+    fn drain_tasks(&self, state: &State) {
```

Comment:
what's the purpose of this?

Reply:
This function drains all of the tasks from the local queue. As the number of runners in the local queue won't change that often I've marked it as cold.